### PR TITLE
Clean up 'Contributing to Rust - Pull Requests'

### LINKED ISSUE #866: Document better that it's OK to use CI for testing
Currently this is documented as
> If your machine is not very powerful, a common strategy is to only use ./x.py check on your local machine and let the CI build test your changes when you push to a PR branch.

But it's hidden all the way down in https://rustc-dev-guide.rust-lang.org/building/prerequisites.html#hardware, which is a long way away from https://rustc-dev-guide.rust-lang.org/contributing.html#pull-requests (which is itself not easy to find).

We should
- [x] Add a note near the 'pull request' section that Rust has plenty of CI capacity for PRs and it's fine to use it
- [ ] Make the pull request section easier to find
- [ ] Link to it from more places

For context, this is meant to be a replacement for https://github.com/rust-lang/rust/pull/76272.

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
Also helps with #866.

### CODE REVIEW

**Comment by LeSeulArtichaut** [comment]
@jyn514 Do you think this is enough to address #866 btw?

File: src/contributing.md [hunk 1]
```diff
@@ -116,32 +90,76 @@ make a documentation change, add
 to the end of the pull request description, and [@rust-highfive][rust-highfive] will assign
 [@steveklabnik][steveklabnik] instead of a random person. This is entirely optional.
 
+In addition to being reviewed by a human, pull requests are automatically tested
+thanks to continuous integration (CI). Basically, every time you open and update
+a pull request, the CI builds the compiler and tests it against the
```
Comment (jyn514) [inline src/contributing.md:95]: nit: CI is a process, not a noun
```suggestion
a pull request, CI builds the compiler and tests it against the
```
or
```suggestion
a pull request, Github Actions builds the compiler and tests it against the
```

File: src/contributing.md [hunk 2]
```diff
@@ -116,32 +90,76 @@ make a documentation change, add
...
+your pull request is in compliance with Rust's style guidelines.
+
+Running continuous integration tests allows PR authors to catch mistakes early
+without going through a first review cycle, and also helps reviewers stay aware
+of the status of a particular pull request.
+
+Rust has plenty of CI capacity, and you should never have to worry about wasting
+computational resources each time you push a change. It is also perfectly fine
+(and even encouraged!) to use the CI to test your changes if it can help your
+productivity, e.g. if your machine is not very powerful.
```
Comment (jyn514) [inline src/contributing.md:106]: This is true even for people with powerful machines ;) rustc is a bit of a hog.
```suggestion
productivity. In particular, we don't recommend running the full `x.py test` suite locally,
since it takes a very long time to execute.
```

File: src/contributing.md [hunk 3]
```diff
@@ -116,32 +90,76 @@ make a documentation change, add
...
+with one another are marked as "always roll up".
 
 [rust-highfive]: https://github.com/rust-highfive
 [steveklabnik]: https://github.com/steveklabnik
 [@bors]: https://github.com/bors
 [merge-queue]: https://buildbot2.rust-lang.org/homu/queue/rust
 
-Speaking of tests, Rust has a comprehensive test suite. More information about
-it can be found [here][rctd].
+### Opening a PR
```
Comment (jyn514) [inline src/contributing.md:132]: Hmm, I think I'd prefer for this to be a separate page to avoid making this one too long. But there's already a lot of sections, so maybe splitting them all up should be a separate PR?

File: src/contributing.md [hunk 4]
```diff
@@ -116,32 +90,76 @@ make a documentation change, add
...
+
+You are now ready to file a pull request? Great! Here are a few points you
+should be aware of.
+
+All pull requests should be filed against the `master` branch, except in very
+particular scenarios. Unless you know for sure that you should target another
+branch, `master` will be the right choice.
```
Comment (jyn514) [inline src/contributing.md:139]:
```suggestion
branch, `master` will be the right choice (it's also the default).
```

File: src/contributing.md [hunk 5]
```diff
@@ -116,32 +90,76 @@ make a documentation change, add
...
+
+Make sure your pull request is in compliance with Rust's style guidelines by running
+
+    $ ./x.py test tidy
```
Comment (jyn514) [inline src/contributing.md:143]: No need to make them run it twice.
```suggestion
    $ ./x.py test tidy --bless
```

File: src/contributing.md [hunk 6]
```diff
@@ -116,32 +90,76 @@ make a documentation change, add
...
+We recommand to make this check before every pull request (and every new commit
+in a pull request); you can add [git hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks)
+before every push to make sure you never forget to make this check. The
+CI will also run tidy and will fail if tidy fails.
+
+Rust follows a _no merge-commit policy_, meaning, when you encounter merge
+conflicts you are expected to always rebase instead of merging.  E.g. always use
+rebase when bringing the latest changes from the master branch to your feature
+branch.  Also, please make sure that fixup commits are squashed into other
+related commits with meaningful commit messages.
```
Comment (jyn514) [inline src/contributing.md:154]: Once someone writes up #850, this should link there. But no need to solve two issues at once ;)

**Comment by jyn514** [comment]
> @jyn514 Do you think this is enough to address #866 btw?

I think it's not discoverable enough still ... I think to mark that as closed we should split up the contributing page into multiple smaller ones, then I'd be happy enough. As it is there's a lot of text to wade through to find what you're looking for.

**Comment by jyn514** [comment]
This is definitely much better than before though!

**Comment by LeSeulArtichaut** [comment]
Changed it to not close #866.

File: src/contributing.md (hunk 1 above)
Comment (LeSeulArtichaut) [reply src/contributing.md:95]: I think I've seen occurences of "the CI" in other parts of the guide... Not sure that's worth fixing, I'll just change that for my PR.

File: src/contributing.md (hunk 3 above)
Comment (LeSeulArtichaut) [reply src/contributing.md:132]: I think I could do that in a separate PR

**Review by jyn514 — APPROVED** [review]
This is somewhat of a big change - maybe someone else wants to take a look? but LGTM

**Review by mark-i-m — APPROVED** [review]
Thanks! LGTM :)

File: src/contributing.md [hunk 7]
```diff
@@ -116,32 +90,77 @@ make a documentation change, add
 to the end of the pull request description, and [@rust-highfive][rust-highfive] will assign
 [@steveklabnik][steveklabnik] instead of a random person. This is entirely optional.
 
+In addition to being reviewed by a human, pull requests are automatically tested
+thanks to continuous integration (CI). Basically, every time you open and update
+a pull request, CI builds the compiler and tests it against the
```
Comment (tshepang) [inline src/contributing.md:95]:
```suggestion
a PR, CI builds the compiler and tests it against the
```

Comment (tshepang) [reply src/contributing.md:95]: This avoids having to decide between *pull request* and *Pull Request*

Comment (LeSeulArtichaut) [reply src/contributing.md:95]: I don't think Pull Request vs pull request is a problem however

Comment (tshepang) [reply src/contributing.md:95]: The inconsistency bothered me

File: src/contributing.md [hunk 8]
```diff
@@ -116,32 +90,77 @@ make a documentation change, add
...
+
+All pull requests should be filed against the `master` branch, except in very
+particular scenarios. Unless you know for sure that you should target another
+branch, `master` will be the right choice (it's also the default).
+
+Make sure your pull request is in compliance with Rust's style guidelines by running
+
+    $ ./x.py test tidy --bless
+
+We recommand to make this check before every pull request (and every new commit
```
Comment (tshepang) [inline src/contributing.md:146]:
```suggestion
We recommend to make this check before every pull request (and every new commit
```
